### PR TITLE
[UnitTests] Apply correct requires_gpu() pytest marks for parametrized target

### PR DESCRIPTION
**Pull request by Lunderberg**
Previously, the addition of tvm.testing._target_to_requirement pytest marks was handled by the parametrize_targets function.  The _auto_parametrize_target function assumed that a unit test that was already parametrized had all markings needed.  If a unit test was explicitly parametrized using @pytest.mark.parametrize, these marks would be missing.

In most cases, this explicit use of @pytest.mark.parametrize('target', ...) should be avoided, but has value in the case of marking with multiple parameters with @pytest.mark.parametrize('target,other', ...).  This use case isn't yet supported by the tvm.testing.parameters function.  Therefore, if this occurs, detect it and add the appropriate marks.

### CODE REVIEW

**Comment by Lunderberg** [comment]
Related to #8387.  This PR removes the `@requires_gpu` annotation, as it is now applied automatically, which should avoid any similar issues in the future.

File: python/tvm/testing.py [hunk 1]
```diff
@@ -795,15 +795,32 @@ def _auto_parametrize_target(metafunc):
...
-        parametrized_args = [
-            arg.strip()
-            for mark in metafunc.definition.iter_markers("parametrize")
-            for arg in mark.args[0].split(",")
-        ]
-
-        if "target" not in parametrized_args:
-            # Check if the function is marked with either excluded or
-            # known failing targets.
+        for mark in metafunc.definition.iter_markers("parametrize"):
```
Comment (comaniac) [inline python/tvm/testing.py:798]: nit: Better not to use loop-else construct (e.g., https://intoli.com/blog/for-else-in-python/).

Comment (Lunderberg) [reply python/tvm/testing.py]: No problem, and I can change that.  The loop-else reads as cleaner to me, even in the blog post's example, but having a consistent style with the rest of the TVM codebase is far more important so I'll avoid using them.

**Comment by Lunderberg** [comment]
Failures in CI were due to ONNX-specific tests.  This PR is now rebased on top of #8574 , which should be merged prior to this PR.

**Review by mbrookhart — CHANGES_REQUESTED** [review]
I don't understand what this is doing, and it's enabled tests that aren't marked to run on GPU, and this freaks me out. I really think we need a full review/doc update/RFC of what on earth the testing parameterization is doing, because I really have no idea how to tell CI where to run or disable certain tests anymore. The fact that we've been silently disabling and enabling tests is terrifying.

**Comment by Lunderberg** [comment]
Agreed, this ended up being a much larger impact than I had expected, and this PR shouldn't be merged until/unless we have consensus on it.  (This comment ended up being a bit longer than I had expected, as I was looking up some history for my own knowledge, and figured I might as well write it down.)

Regarding changes to how unit tests are parametrized, when they are marked with `@pytest.mark.gpu`, here's the history as best as I know.

1. (State prior to listed changes.) Unit tests explicitly apply to a particular target.  The CI explicitly checks for the presence of `@pytest.mark.gpu`, from the [`-m` gpu` argument](https://github.com/apache/tvm/blob/main/tests/scripts/task_python_unittest_gpuonly.sh#L20).  A test can run over multiple targets by explicitly looping over `tvm.testing.enabled_targets()`.
2. #6331 added the `@tvm.testing.parametrize_targets` decorator, to parametrize across all enabled targets.  A unit-test decorated with `parametrize_targets` has an independent test run for each item in `tvm.testing.enabled_targets()`.  These independent tests are marked with the appropriate marks (e.g. `tvm.testing.requires_cuda`, which includes `pytest.mark.gpu`) based on the target in use.
3. #8010 added parametrized fixtures for `target` and `dev`.  The goal was to reduce the amount of boilerplate needed for any given unit test.  A unit test that accepts arguments of `target` or `dev` is automatically parametrized over the enabled targets.  This behavior is described in the [tvm-rfc repo, PR#7](https://github.com/apache/tvm-rfcs/pull/7).  There's some conversation on finalizing exactly how and what should be cached, but the RFC is up-to-date with the latest conversation and implementation.  Relevant to the GPU-marking, this followed the same convention as #6331, marking tests as requiring a GPU if the target parameter is a GPU target.

This worked, but allowed for accidentally forgetting an annotation for a target-specific test.  This particular PR was motivated by #8387 , which added a missing annotation of `@requires_gpu` to a test that explicitly used `pytest.mark.parametrize`.  In order to remove that as a potential failure mode, this PR identifies cases where the test has been explicitly parametrized over targets, and adds the same marks as if it had used `@tvm.testing.parametrize_targets`, including `@tvm.testing.requires_gpu` if it is a GPU target.  That is why the ONNX-specific tests started running with this PR, because the parametrization of `target` then received the target-specific marks.

I think the source of my misunderstanding was that `@tvm.testing.requires_gpu` serves two purposes.  The first is to indicate that the test should be skipped if no CPU is present.  The second (which I hadn't known) is to indicate that a test should be run on the gpu-nodes at all.  By adding the target-specific marks, I had been assuming that their lack would be accidental rather than intentional.

**Comment by Lunderberg** [comment]
With the parametrized targets, we can explicitly mark the test with `@tvm.testing.excluded_targets('cuda')` or `@tvm.testing.known_failing_targets('cuda')` to remove parametrization over those specific tests.

**Comment by Lunderberg** [comment]
Added the ONNX cleanup and use of `tvm.testing.known_failing_targets` for onnx frontend tests to this PR, after rebasing on top of #8621.  They both needed to adjust the `target` parameter, and this PR already enables the functionality needed.

@mbrookhart

**Review by mbrookhart — APPROVED** [review]
LGTM. Agreed that we'll follow up with a user-facing doc on pytest usage. Thanks, @Lunderberg!

**Comment by Lunderberg** [comment]
Current failure is due to the automatic-marking expected a string, and not a `tvm.target.Target` object.  This particular issue is resolved with some refactoring that went alongside #8383.  Will cherry-pick that commit for the refactoring, leaving the cudnn-enabling by default on that PR.

**Comment by Lunderberg** [comment]
Bumping to restart CI after #8656 .

**Pull request merged**